victim-loss-identifier/src/fetcher.ts

```diff
@@ -0,0 +1,263 @@
+      return this.ethPriceCache.get(timestamp);
+    }
+
+    const maxTries = 3;
+    let tries = 0;
+
+    while (tries < maxTries) {
+      try {
+        const url = `https://coins.llama.fi/prices/historical/${timestamp}/coingecko:ethereum`;
+        const response = await fetch(url);
+        const data = (await response.json()) as coinData;
+        const price = data.coins["coingecko:ethereum"]?.price;
+
+        if (price == null) {
+          tries++; // Increment tries counter for null price
+          console.log(`Fetched price is null (attempt ${tries})`);
+        } else {
+          this.ethPriceCache.set(timestamp, price);
+          return price;
+        }
+      } catch (e) {
+        tries++;
+        console.log(`Error in fetching ETH price (attempt ${tries}):`, e);
+        if (tries === maxTries) {
+          throw e;
+        }
+      }
+
+      // Wait for 1 second before retrying
+      await new Promise((resolve) => setTimeout(resolve, 1000));
+    }
+
+    return undefined;
+  };
+
+  private getFloorPriceInEth = async (nftAddress: string): Promise<number> => {
+    const maxTries = 3;
+    let tries = 0;
+
+    while (tries < maxTries) {
+      try {
+        const floorPrice: GetFloorPriceResponse =
+          await this.alchemy.nft.getFloorPrice(nftAddress);
+
+        const latestEntry = Object.values(floorPrice).reduce(
+          (latest: FloorPriceMarketplace | null, current) => {
+            if ("error" in current) {
+              return latest;
+            }
+
+            if (
+              !latest ||
+              new Date(current.retrievedAt) > new Date(latest.retrievedAt)
+            ) {
+              return current;
+            }
+            return latest;
+          },
+          {
+            floorPrice: 0,
+            priceCurrency: "",
+            collectionUrl: "",
+            retrievedAt: "2000-01-01T00:00:00.000Z",
+          }
+        );
+
+        if (latestEntry.priceCurrency !== "ETH") {
+          return 0;
+        }
+
+        return latestEntry.floorPrice;
+      } catch (e) {
+        tries++;
+        console.error(`Error fetching floor price (attempt ${tries}):`, e);
+        if (tries === maxTries) {
+          throw e;
+        }
+
+        // Wait for 1 second before retrying
+        await new Promise((resolve) => setTimeout(resolve, 1000));
+      }
+    }
+
+    return 0;
+  };
+
+  getNftCollectionFloorPrice = async (
+    nftAddress: string,
+    timestamp: number
+  ): Promise<number> => {
+    if (this.floorPriceCache.has(timestamp)) {
+      return this.floorPriceCache.get(timestamp)!;
+    }
+
+    const ethPrice = await this.getNativeTokenPrice(timestamp);
+    const floorPriceInEth = await this.getFloorPriceInEth(nftAddress);
+
+    if (!ethPrice || !floorPriceInEth) {
+      return 0;
+    }
+
+    const floorPrice = floorPriceInEth * ethPrice;
+    this.floorPriceCache.set(timestamp, floorPrice);
```

Comment:
This line is setting the floor price to the "key" of `timestamp`. However, setting the price in this way doesn't utilize the NFT collection address. So it runs the risk of using the floor price of one NFT collection at one timestamp for an unrelated NFT collection, if the same timestamp were to be used.

Ideally, we should incorporate the address as well. Perhaps concat'ing the timestamp and address into a single key? (If wanting to keep the key a `number`, we could convert the hexadecimal value of the address to decimal? Or convert the key to a `string` type?)

Reply:
Of course! Changed it to string type